macOS: Enumerate GPUs first; prefer low-power non-removable; fall back to system default

cacaosteve opened the pull request:
Problem: Some macOS environments report no devices via MTLCopyAllDevices, causing startup failure with “unable to access a compatible graphics device,” especially on Apple Silicon.
Change: Prefer MTLCreateSystemDefaultDevice (metal::Device::system_default()) first. If None, enumerate devices and select a non‑removable, low‑power device by preference.
Why this works: On Apple Silicon the system default is the unified GPU; on Intel, the fallback keeps a stable policy and avoids accidentally picking removable/high‑power devices.
Impact: Fixes startup on affected ASi systems; improves selection consistency on Intel multi‑GPU. Behavior unchanged where system_default() succeeds.
Risk: Low. Aligns with Apple’s recommended selection path. Still fails early with a clearer message if no Metal devices exist.

Closes #37689.

Release Notes:
- Fixed: Startup failure on some Apple Silicon machines when Metal device enumeration returned no devices by falling back to the system default device.

cacaosteve commented:
recheck

filipwiech commented on crates/gpui/src/platform/mac/metal_renderer.rs:
```diff
@@ -132,11 +132,18 @@ impl MetalRenderer {
+                std::process::exit(1);
+            }
+            // When multiple devices exist (e.g., Intel iGPU + dGPU or eGPU), prefer low-power, non-removable.
+            devices.sort_by_key(|d| (d.is_removable(), !d.is_low_power()));
```
The original code sorted by `(is_removable(), is_low_power())`, while this new one sorts by `(is_removable(), !is_low_power()`: was there a bug before? :thinking:

cacaosteve committed:
mac/metal: prefer system_default() device; fallback to enumeration

Use metal::Device::system_default() first to select the GPU.
Fall back to Device::all() only if default device is unavailable.
When enumerating, prefer non-removable, low-power devices for stability.
Improves startup on ASi where enumeration returns empty; keeps sensible default on Intel multi‑GPU.
Clarify error when no Metal devices are found.
Fixes #37689

JunkuiZhang commented:
From my perspective, this PR has two issues:

- When enumerating GPUs, low-power ones should come first. That way on Intel Macs we’ll use the iGPU instead of the dGPU.

- On Intel Macs, `metal::Device::system_default()` actually always returns the dGPU, which means the code for selecting another GPU is never reached, so this PR will instroduce a regression.

cacaosteve committed:
prefer low-power, non-removable GPU; use system_default only for 0/1 devices

JunkuiZhang commented:
I think your changes make things a bit more complicated than they need to be. If I understand correctly, the issue on your machine is that `devices.pop()` returns `None`, which leads to an error being logged and then an immediate exit. A better approach, in my opinion, would be to log the error but not exit right away. Instead, fall back to `metal::Device::system_default()`.

cacaosteve committed:
mac/metal: enumerate GPUs first, prefer low-power non-removable; fallback to system_default

cacaosteve commented:
recheck

cacaosteve committed:
fmt: apply rustfmt

JunkuiZhang committed:
Simplify everything

JunkuiZhang committed:
Update comments

JunkuiZhang commented:
I pushed a few changes to your PR, could you check if everything still runs fine on your machine?

JunkuiZhang commented:
The change you suggested seems logically the same as mine, the only difference is that you’re using `!d.is_low_power()` while I’m using `d.is_low_power()`. From my testing, `d.is_low_power()` seems to be the correct one. But maybe I misunderstood your point, happy to adjust if that’s the case.

cacaosteve committed:
macOS/metal: Enumerate GPUs; Prefer iGPU via min_by_key; Fallback to system_default

cacaosteve commented:
I cleaned it up a bit more, I hope that's okay. My testing on MacBook Pro M1 Max always works, I was not able to reproduce the original issue.

Prefer non-removable, low-power GPU (iGPU) by ordering with min_by_key(|d| (d.is_removable(), !d.is_low_power())), yielding iGPU → dGPU → eGPU.
If enumeration returns empty (as seen in #37689), log and fall back to metal::Device::system_default(); exit only if both paths fail.
Fixes the prior sort+pop behavior, which could select an eGPU/dGPU unintentionally and is harder to reason about.
Avoids Intel regression where system_default() often picks the dGPU; Apple Silicon (single GPU) behavior is unchanged.

localcc committed:
retrigger ci
